### PR TITLE
Fix/several small fixes

### LINKED ISSUE #2124: [Discord Suggestion] Bed Coloring/Choice
Can it be made that the builder will use any bed instead of just the white ones? When I picked up my citizen's huts the beds were red and couldn't be used again.

From: Zippi#4573

### LINKED ISSUE #2497: Buider converts white beds to red beds
### Minecolonies version 1.12.2-0.8.7582

version X.X.XXXX or curse X.X.XX

### Expected behavior

- The builder will ask for a white bed and place a white bed. When you upgrade the building he will knock up the white bed and reuse it for the upgrade. 

### Actual behaviour

- The builder will ask for a white bed but when he places it, it gets turned into a red bed. When you upgrade the building he will knock up the red bed and ignore it (leaving it floating) and then request more white beds for the upgrade. I am ending with lots of spare red beds as a result.

### Steps to reproduce the problem

1. If the builder asks for a white bed during a build observe when he places it down
2. The bed will get converted into a red bed when he places it down.

### LINKED ISSUE #2549: colonyX.dat files excessively large
### Minecolonies version

version X.X.9540 but could be a carry over form previous versions

### Expected behavior

- colonyX.dat file remains a normal size

### Actual behaviour

- colonyX.dat file is HUGE

### Steps to reproduce the problem

1. Not sure how to reproduce but we have 2 very large files.

Colony 2 - My colony: 17 Level 5 buildings. Citzens restricted to 5. (2 builders, 1 dman, 1 cowboy, 1 miner) But I was at 10 without issue. colony2.dat 194kb,  looking at file using NBTEdit: workorder entries 171 entries (which still seems excessive)

Colony 1 - Our problematic colony (10 Level 2 buildngs. 5 Citizens) colony1.dat 3717 kb. Workorder entries: 13833   (This is the colony where our TPS drops to 4 every 20-30 seconds if he has a builder working)

Colony1.dat (renamed 1-1 for editing purposes)
screenshot 2928

Colony 5 - Player doesn't seem to have problems or cause server lag, but he really doesn't build his colony much (3 buildings Level 1, 4 citizens) colony5.dat 2135kb. Workorder entries: 8713

Creating this Issue from work, but tonight I'm going to try deleting the 13k entries from colony1.dat and see if server performance is back to normal when that player has a builder working. Will report late tonight with results.

### PR DESCRIPTION
**Pull request by Raycoms**
Closes #2124
Closes #2497
Closes #2549

# Changes proposed in this pull request:
- Builder now requests the correct color of beds and also places the correct color.
- Miner is now more efficient mining horizontally.
- Miner now closes all his open workorders. (Should reduce CPU and RAM load)

Review please

### CODE REVIEW

File: src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingMiner.java [hunk 1]
```diff
@@ -41,152 +40,79 @@
  */
 public class BuildingMiner extends AbstractBuildingStructureBuilder
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingMiner.java:41]: MINOR Override the "equals" method in this class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2160)

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/Node.java [hunk 2]
```diff
@@ -273,14 +281,74 @@ public Vec2i getWestNodeCenter()
...
 
+    /**
+     * Return a random next node to work at, might be at this node or at a parent.
+     * @return the next node to go to.
+     * @param level the level it is part of.
+     */
+    @Nullable
+    public Node getRandomNextNode(final Level level, final int step)
+    {
+        if (step > 3)
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/Node.java:292]: MAJOR Assign this magic number 3 to a well-named constant, and use the constant instead. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/Node.java [hunk 3]
```diff
@@ -273,14 +281,74 @@ public Vec2i getWestNodeCenter()
...
+        {
+            return null;
+        }
+
+        final Node nextNode;
+        switch (random.nextInt(3))
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/Node.java:298]: MAJOR Assign this magic number 3 to a well-named constant, and use the constant instead. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/Node.java [hunk 4]
```diff
@@ -273,14 +281,74 @@ public Vec2i getWestNodeCenter()
...
+            case 2:
+                nextNode = level.getOpenNode(getEastNodeCenter());
+                break;
+            default:
+                nextNode = level.getOpenNode(getWestNodeCenter());
+        }
+
+        if (nextNode == null || nextNode.style == NodeType.SHAFT)
+        {
+            final Node parent = level.getOpenNode(getParent());
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/Node.java:315]: MAJOR Rename "parent" which hides the field declared at line 55. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AHiddenFieldCheck)

File: src/main/java/com/minecolonies/coremod/placementhandlers/PlacementHandlers.java [hunk 5]
```diff
@@ -123,8 +125,8 @@ public static boolean checkForListInInvAndRequest(@NotNull final AbstractEntityA
     {
         @Override
         public Object handle(
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/placementhandlers/PlacementHandlers.java:127]: MAJOR Refactor this method to reduce its Cognitive Complexity from 23 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/EntityAIStructureMiner.java [hunk 6]
```diff
@@ -797,10 +821,24 @@ public void executeSpecificCompleteActions()
...
         //Send out update to client
         getOwnBuilding().markDirty();
-
-        job.setStructure(null);
+         job.setStructure(null);
+
+         final Colony colony = worker.getCitizenColonyHandler().getColony();
+         if (colony != null)
+         {
+             final List<WorkOrderBuildMiner> workOrders = colony.getWorkManager().getWorkOrdersOfType(WorkOrderBuildMiner.class);
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/EntityAIStructureMiner.java]: MINOR Make this line start at column 13. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AIndentationCheck)

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/EntityAIStructureMiner.java [hunk 7]
```diff
@@ -797,10 +821,24 @@ public void executeSpecificCompleteActions()
...
+             if (workOrders.size() > 5)
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/EntityAIStructureMiner.java]: MAJOR Assign this magic number 5 to a well-named constant, and use the constant instead. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/EntityAIStructureMiner.java [hunk 8]
```diff
@@ -540,9 +547,16 @@ private AIState executeNodeMining()
 
     private AIState searchANodeToMine(@NotNull final Level currentLevel)
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/EntityAIStructureMiner.java:548]: MAJOR The Cyclomatic Complexity of this method "searchANodeToMine" is 17 which is greater than 15 authorized. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/EntityAIStructureMiner.java [hunk 9]
```diff
@@ -797,10 +821,24 @@ public void executeSpecificCompleteActions()
...
-        job.setStructure(null);
+         job.setStructure(null);
+
+         final Colony colony = worker.getCitizenColonyHandler().getColony();
+         if (colony != null)
+         {
+             final List<WorkOrderBuildMiner> workOrders = colony.getWorkManager().getWorkOrdersOfType(WorkOrderBuildMiner.class);
+             if (workOrders.size() > 5)
+             {
+                 for (WorkOrderBuildMiner order : workOrders)
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/EntityAIStructureMiner.java]: MINOR Make this line start at column 17. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AIndentationCheck)

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/EntityAIStructureMiner.java [hunk 10]
```diff
@@ -797,10 +821,24 @@ public void executeSpecificCompleteActions()
...
+                 {
+                     colony.getWorkManager().removeWorkOrder(order.getID());
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/EntityAIStructureMiner.java]: MINOR Make this line start at column 21. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AIndentationCheck)

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/EntityAIStructureMiner.java [hunk 11]
```diff
@@ -567,6 +581,14 @@ else if (vectorZ == 1)
             rotation = ROTATE_ONCE;
         }
 
+        final Node parentNode = currentLevel.getNode(workingNode.getParent());
+        if (parentNode.getStyle() != Node.NodeType.SHAFT && (parentNode.getStatus() != Node.NodeStatus.COMPLETED || world.getBlockState(new BlockPos(parentNode.getX(), currentLevel.getDepth() + 2, parentNode.getZ())).getBlock() != Blocks.AIR))
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/EntityAIStructureMiner.java:585]: MAJOR Split this 243 characters long line (which is greater than 180 authorized). [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS00103)

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/EntityAIStructureMiner.java [hunk 12]
```diff
@@ -797,10 +821,24 @@ public void executeSpecificCompleteActions()
             minerBuilding.setCurrentLevel(minerBuilding.getNumberOfLevels());
             minerBuilding.resetStartingLevelShaft();
         }
+        super.executeSpecificCompleteActions();
+
         //Send out update to client
         getOwnBuilding().markDirty();
-
-        job.setStructure(null);
+         job.setStructure(null);
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/EntityAIStructureMiner.java]: MINOR Make this line start at column 9. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AIndentationCheck)

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/EntityAIStructureMiner.java [hunk 13]
```diff
@@ -10,6 +11,8 @@
 import com.minecolonies.coremod.entity.ai.basic.AbstractEntityAIStructureWithWorkOrder;
 import com.minecolonies.coremod.entity.ai.util.AIState;
 import com.minecolonies.coremod.entity.ai.util.AITarget;
+import com.minecolonies.coremod.entity.pathfinding.PathResult;
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/EntityAIStructureMiner.java:14]: MINOR Remove this unused import 'com.minecolonies.coremod.entity.pathfinding.PathResult'. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AUselessImportCheck)

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/EntityAIStructureMiner.java [hunk 14]
```diff
@@ -10,6 +11,8 @@
...
+import com.minecolonies.coremod.inventory.GuiHandler;
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/EntityAIStructureMiner.java:15]: MINOR Remove this unused import 'com.minecolonies.coremod.inventory.GuiHandler'. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by BuildTeamLinter** [comment]
SonarQube analysis reported 98 issues
* ![MAJOR][MAJOR] 86 major
* ![MINOR][MINOR] 10 minor
* ![INFO][INFO] 2 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [blocks#](https://github.com/ldtteam/minecolonies/blob/8b0f6cc6bbe4f700f93edcfb3bf84e46ce9c3f31/src/main/java/com/minecolonies/coremod/blocks#): Add a 'package-info.java' file to document the 'blocks' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [cactus#](https://github.com/ldtteam/minecolonies/blob/8b0f6cc6bbe4f700f93edcfb3bf84e46ce9c3f31/src/main/java/com/minecolonies/coremod/blocks/cactus#): Add a 'package-info.java' file to document the 'cactus' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [decorative#](https://github.com/ldtteam/minecolonies/blob/8b0f6cc6bbe4f700f93edcfb3bf84e46ce9c3f31/src/main/java/com/minecolonies/coremod/blocks/decorative#): Add a 'package-info.java' file to document the 'decorative' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [huts#](https://github.com/ldtteam/minecolonies/blob/8b0f6cc6bbe4f700f93edcfb3bf84e46ce9c3f31/src/main/java/com/minecolonies/coremod/blocks/huts#): Add a 'package-info.java' file to document the 'huts' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [interfaces#](https://github.com/ldtteam/minecolonies/blob/8b0f6cc6bbe4f700f93edcfb3bf84e46ce9c3f31/src/main/java/com/minecolonies/coremod/blocks/interfaces#): Add a 'package-info.java' file to document the 'interfaces' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [schematic#](https://github.com/ldtteam/minecolonies/blob/8b0f6cc6bbe4f700f93edcfb3bf84e46ce9c3f31/src/main/java/com/minecolonies/coremod/blocks/schematic#): Add a 'package-info.java' file to document the 'schematic' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [types#](https://github.com/ldtteam/minecolonies/blob/8b0f6cc6bbe4f700f93edcfb3bf84e46ce9c3f31/src/main/java/com/minecolonies/coremod/blocks/types#): Add a 'package-info.java' file to document the 'types' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [gui#](https://github.com/ldtteam/minecolonies/blob/8b0f6cc6bbe4f700f93edcfb3bf84e46ce9c3f31/src/main/java/com/minecolonies/coremod/client/gui#): Add a 'package-info.java' file to document the 'gui' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [render#](https://github.com/ldtteam/minecolonies/blob/8b0f6cc6bbe4f700f93edcfb3bf84e46ce9c3f31/src/main/java/com/minecolonies/coremod/client/render#): Add a 'package-info.java' file to document the 'render' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [BuildingMiner.java#L391](https://github.com/ldtteam/minecolonies/blob/8b0f6cc6bbe4f700f93edcfb3bf84e46ce9c3f31/src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingMiner.java#L391): Assign this magic number 3 to a well-named constant, and use the constant instead. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Review by marchermans — APPROVED** [review]

**Comment by BuildTeamLinter** [comment]
SonarQube analysis reported 98 issues
* ![MAJOR][MAJOR] 86 major
* ![MINOR][MINOR] 10 minor
* ![INFO][INFO] 2 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [blocks#](https://github.com/ldtteam/minecolonies/blob/b11f037a7801c7088031af4547b7cb7430d0b37e/src/main/java/com/minecolonies/coremod/blocks#): Add a 'package-info.java' file to document the 'blocks' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [cactus#](https://github.com/ldtteam/minecolonies/blob/b11f037a7801c7088031af4547b7cb7430d0b37e/src/main/java/com/minecolonies/coremod/blocks/cactus#): Add a 'package-info.java' file to document the 'cactus' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [decorative#](https://github.com/ldtteam/minecolonies/blob/b11f037a7801c7088031af4547b7cb7430d0b37e/src/main/java/com/minecolonies/coremod/blocks/decorative#): Add a 'package-info.java' file to document the 'decorative' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [huts#](https://github.com/ldtteam/minecolonies/blob/b11f037a7801c7088031af4547b7cb7430d0b37e/src/main/java/com/minecolonies/coremod/blocks/huts#): Add a 'package-info.java' file to document the 'huts' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [interfaces#](https://github.com/ldtteam/minecolonies/blob/b11f037a7801c7088031af4547b7cb7430d0b37e/src/main/java/com/minecolonies/coremod/blocks/interfaces#): Add a 'package-info.java' file to document the 'interfaces' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [schematic#](https://github.com/ldtteam/minecolonies/blob/b11f037a7801c7088031af4547b7cb7430d0b37e/src/main/java/com/minecolonies/coremod/blocks/schematic#): Add a 'package-info.java' file to document the 'schematic' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [types#](https://github.com/ldtteam/minecolonies/blob/b11f037a7801c7088031af4547b7cb7430d0b37e/src/main/java/com/minecolonies/coremod/blocks/types#): Add a 'package-info.java' file to document the 'types' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [gui#](https://github.com/ldtteam/minecolonies/blob/b11f037a7801c7088031af4547b7cb7430d0b37e/src/main/java/com/minecolonies/coremod/client/gui#): Add a 'package-info.java' file to document the 'gui' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [render#](https://github.com/ldtteam/minecolonies/blob/b11f037a7801c7088031af4547b7cb7430d0b37e/src/main/java/com/minecolonies/coremod/client/render#): Add a 'package-info.java' file to document the 'render' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [BuildingMiner.java#L391](https://github.com/ldtteam/minecolonies/blob/b11f037a7801c7088031af4547b7cb7430d0b37e/src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingMiner.java#L391): Assign this magic number 3 to a well-named constant, and use the constant instead. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**